StefanVelkoski/semaphore-noir
# Siosw/milestone one

### Milestone 1

## Description

The Semaphore project’s circuits written in Noir according to original specs, alongside tests with off-chain proof verification

Offchain proof verification will consist of a noir semaphore circuit and modified Group, and Identity classes forked from the Semaphore EF project. As we will use Pedersen hashes instead of Poseidon hashes, the forked classes must be modified accordingly.
The main deliverable will be an integration test showcasing:
Creating a group
Proving group membership in ZK


## What does this this implement?

This PR implements a noir semaphore circuit with modified Group and Identity classes/packages to be compatible with the forked Semaphore EF project. With this implementation one can create a group and prove a group membership using ZK technology.

## What are the differences between the original Semaphore implementation and ours?

We are using the Noir language for the circuits and the following libraries for circuit to js communication.
`  "@noir-lang/aztec_backend": "0.2.0",
    "@noir-lang/barretenberg": "github:noir-lang/barretenberg#39a1547875f941ef6640217a42d8f34972425c97",
    "@noir-lang/noir_wasm": "0.3.2-fa0e9cff",
`
For the hash function we are implementing the Pedersen hash function while Poseidon is used on the original implementation.

## How to test the changes?

To successfully test the implementation run the following commands:

Clone the repo:
`git clone git@github.com:StefanVelkoski/semaphore-noir.git`

Install dependencies:
`cd semaphore && yarn`

Run the test suites:
`yarn test`

Note that for the implementation of this PR, the relevant test are the `group.test.ts`, `identity.test.ts` and the most important  - `offchain.test.ts`

We encourage checking out the tests and playing around with the creation of Groups and Identities.

## Any other comments?

The current circuit implementation will have some minor changes  as soon as some of the needed features are implemented in Noir. This is for the purpose of achieving absolute compatibility with the original Semaphore implementation.

The Nargo CLI implementation of the `offchain.test.ts` will also be a subject to change to a much more simple typescript implementation for compiling, proving and verifying once it is implemented in the specific libraries.
